### ISSUE TITLE
[PROPOSAL] Delete unmanaged indextemplates and ISM policies

### ISSUE THREAD

**Issue by cthtrifork** [opensearch-project/opensearch-k8s-operator#869]
### What are you proposing?
All unmanaged index templates and ISM policies should be deleted when reconciling the resources by the operator.

### What problems are you trying to solve?
This is to enforce a strict GitOPs approach for these custom resources and is also a convenient way to make it possible for the operator to delete old templates and policies (by omission).

#### Are there any security considerations? 
We will need to create a blacklist for an internal/system templates or policies that should not be affected by this.
It should also be configurable and default this behaviour should not be enabled.

### What will it take to execute?
The operator should be expanded with an option to verify all custom resources against the list in the opensearch and delete unmatched items.

We would be willing to make a PR for this feature, if is no major concerns.
We would add it to the reconciler logic for ISM policies and templates.

**Comment by rkthtrifork** [comment]
It seems there is
```go
podList := &v1.ResourceList{}
err := r.List(ctx, podList, opts...)
```
which I think gets all the resources and we can filter them. I don't know if it would be reasonable to do it in the reconciler for OpensearchCluster or if it should be in a separate reconciler. Just from a quick look at what seems possible

https://sdk.operatorframework.io/docs/building-operators/golang/references/client/#list

**Comment by swoehrl-mw** [comment]
@rkthtrifork Yes, the list operation is probably the best way to get an overview over all policies currently defined in K8s.

> I don't know if it would be reasonable to do it in the reconciler for OpensearchCluster or if it should be in a separate reconciler.

IMO it should be its own reconciler that is called in the opensearchClusterController (just like for example the snapshot repository reconciler). That way it is a clear separation of concerns and better modularization of the code.

**Comment by synhershko** [comment]
Hey, thanks for opening this and sorry for the wait.

We've made the decision to narrow the operator's scope down to core cluster lifecycle management going forward (provisioning, scaling, upgrades, TLS, etc.) and are deprecating the CRDs that manage OpenSearch-side config like ISM policies, index/component templates, snapshot policies, users, roles, role bindings, tenants and action groups. They're marked deprecated as of #1520 and will be removed in v4.

The reasoning: this is config that's better owned by tools built for declarative OpenSearch config management (e.g. Terraform, or the OpenSearch API/Dashboards directly), and maintaining full parity with these OpenSearch features inside the operator was spreading us too thin on the core stuff people actually depend on.

Given that, we won't be fixing or building out this behavior further, so I'm closing this as won't-fix. If you're relying on this CRD today, it'll keep working until v4, but we'd recommend planning a migration off it. Sorry this isn't the outcome you were hoping for, and thanks again for the report/contribution.
